Kernel minimization

kaczmarczyck opened the pull request:
This PR adds a patch to tock to remove 4 unused capsules for all boards we deploy to:

- `ble_advertising_driver::BLE`
- `ieee802154::RadioDriver`
- `temperature::TemperatureSensor`
- `nonvolatile_storage_driver::NonvolatileStorage` (was only in dev kit)

This reduces the binary size of the kernel significantly:
| board   | before | after |
|---------|--------|-------|
| dev kit | 139 kB | 98 kB |
| dongle  | 131 kB | 98 kB |
| mdk     | 126 kB | 94 kB |

kaczmarczyck committed:
adds patch to remove unused kernel components

jmichelp commented:
I think at this point, it would make more sense to create our own boards and leave the ones from Tock upstream untouched.
We already have a `boards/` directory for the Makerdiary dongle.
It's a much bigger PR but it also reduces at least one patch (for the storage).

kaczmarczyck commented:
So you want the all changes of `boards/nordic/` in patches to be moved / merged into one that creates new boards in `boards/nordic`, like an `opensk_nrf52840dk` etc? This would be a bigger patch refactoring. Might also mean that changes to persistent storage or USB need to update more different patches than before.
@jmichelp

kaczmarczyck committed:
new boards, rebase patches to match

kaczmarczyck committed:
rename boards for deployment

kaczmarczyck commented:
Okay, I added our own 2 boards as copies from the dk and dongle. This PR is now pretty big, but compared to the first commit, doesn't do that much:

1. The boards `nrf52840dk` and `nrf52840_dongle` are copied and added to nordic/boards.
2. The (old) patch 03 and the (new) patch 08 are applied on top of these copied boards.
3. All other patches are kept as-is and applied to the new boards.
4. The deploy script, tests and documentation is changed to use boards `nrf52840dk_opensk` and `nrf52840_dongle_opensk`.

jmichelp commented on boards/nordic/nrf52840dk_opensk/src/main.rs:
```diff
@@ -0,0 +1,396 @@
+            capsules::led::DRIVER_NUM => f(Some(self.led)),
+            capsules::button::DRIVER_NUM => f(Some(self.button)),
+            capsules::rng::DRIVER_NUM => f(Some(self.rng)),
+            capsules::analog_comparator::DRIVER_NUM => f(Some(self.analog_comparator)),
```
non blocking comment: we don't use this neither. Should you need to shrink even more, you can take it away

jmichelp commented on boards/nordic/nrf52840dk_opensk/src/main.rs:
```diff
@@ -0,0 +1,396 @@
+
+    let rng = components::rng::RngComponent::new(board_kernel, &nrf52840::trng::TRNG).finalize(());
+
+    nrf52840::spi::SPIM0.configure(
```
SPI and and analog_comparator could also go away if you need to shrink even more. And this also applies to other boards (but on the dongle, SPI should already be gone).

kaczmarczyck replied:
Okay, will do when required.

jmichelp commented on patches/tock/01-persistent-storage.patch:
```diff
@@ -1,8 +1,8 @@
-diff --git a/boards/nordic/nrf52840_dongle/src/main.rs b/boards/nordic/nrf52840_dongle/src/main.rs
```
This patch could be heavily simplified because only the last chunk needs to be applied on the stock kernel.
It could be pre-applied on the `boards/` directory with this PR.
But let's keep it for a future PR.

kaczmarczyck replied:
Got it.
